src/Microsoft.AspNet.OData.Shared/Builder/LinkGenerationHelpers.cs

```diff
@@ -95,6 +95,99 @@ public static Uri GenerateSelfLink(this ResourceContext resourceContext, bool in
             return new Uri(link);
         }
 
+        /// <summary>
+        /// Generates a contained navigation link following the OData URL conventions for the entity represented by <paramref name="resourceContext"/> and the given
+        /// navigation property.
+        /// </summary>
+        /// <param name="resourceContext">The <see cref="ResourceContext"/> representing the entity for which the navigation link needs to be generated.</param>
+        /// <param name="navigationProperty">The EDM navigation property.</param>
+        /// <param name="includeCast">Represents whether the generated link should have a cast segment representing a type cast.</param>
+        /// <returns>The navigation link following the OData URL conventions.</returns>
+        public static Uri GenerateContainedNavigationPropertyLink(this ResourceContext resourceContext,
+            IEdmNavigationProperty navigationProperty, bool includeCast)
```

Comment:
>includeCast [](start = 60, length = 11)

how does the caller know whether to includeCast? Shouldn't we determine this based on whether the navigationProperty is defined on the same type as the navigationSource?  If so, we don't need the cast, if not we have to include it.